File: src/main/java/mod/yourname/yourmodid/chain/ChainItem.java
package mod.yourname.yourmodid.chain;

import com.simibubi.create.content.contraptions.relays.elementary.CogWheelBlock;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtUtils;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;

import java.lang.reflect.Array;
import java.util.List;

public class ChainItem extends Item {
    public ChainItem(Properties properties) {
        super(properties);
    }

    @Override
    public InteractionResult useOn(UseOnContext pContext) {
        BlockPos clicked = pContext.getClickedPos();
        Level level = pContext.getLevel();
        Player player = pContext.getPlayer();
        Block block = level.getBlockState(clicked).getBlock();

        CompoundTag tag = pContext.getItemInHand().getOrCreateTag();

        if (!(block instanceof CogWheelBlock)) {
            level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.AMETHYST_BLOCK_FALL, SoundSource.PLAYERS, 0.2F, 0.5F);
            return InteractionResult.FAIL;
        }

        if (tag.contains("1")) {
            for (String key : tag.getAllKeys()) {
                if (key.matches(".*\\d.*")) {
                    BlockPos currSel = NbtUtils.readBlockPos(tag.getCompound(key));

                    if (currSel.equals(clicked)) {
                        player.getCooldowns().addCooldown(this, 20);
                        level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.CHAIN_BREAK, SoundSource.PLAYERS, 1.0F, 1.0F);
                        tag.remove("First");
                        return InteractionResult.FAIL;
                    }
                }
            }
            if (player.isShiftKeyDown()) {
                BlockPos currSel = NbtUtils.readBlockPos(tag.getCompound("1"));
                System.out.println(currSel + "Is Connected to" + clicked);
                player.getCooldowns().addCooldown(this, 10);
                level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.CHAIN_PLACE, SoundSource.PLAYERS, 1.0F, 1.0F);
                return InteractionResult.CONSUME;
            }
            Integer last = tag.getAllKeys().size();
            tag.put((last + 1).toString(), NbtUtils.writeBlockPos(clicked))
            pContext.getItemInHand().setTag(tag);
            player.getCooldowns().addCooldown(this, 10);

            level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.IRON_TRAPDOOR_OPEN, SoundSource.PLAYERS, 1.0F, 1.0F);
            return InteractionResult.SUCCESS;
        }

        tag.put("1", NbtUtils.writeBlockPos(clicked));
        pContext.getItemInHand().setTag(tag);
        player.getCooldowns().addCooldown(this, 10);

        level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.IRON_TRAPDOOR_CLOSE, SoundSource.PLAYERS, 1.0F, 1.0F);
        return InteractionResult.SUCCESS;
    }
}

File: src/main/kotlin/mod/yourname/yourmodid/kinetic/chain/ChainItem.kt
package mod.yourname.yourmodid.kinetic.chain

import com.simibubi.create.content.contraptions.relays.elementary.CogWheelBlock
import com.simibubi.create.repack.registrate.util.nullness.NonNullFunction
import net.minecraft.nbt.NbtUtils
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import net.minecraft.world.InteractionResult
import net.minecraft.world.item.Item
import net.minecraft.world.item.context.UseOnContext

class ChainItem(pProperties: Properties) : Item(pProperties) {
    override fun useOn(pContext: UseOnContext): InteractionResult {
        val clicked = pContext.clickedPos
        val level = pContext.level
        val player = pContext.player
        val block = level.getBlockState(clicked).block
        val tag = pContext.itemInHand.getOrCreateTag()
        if (block !is CogWheelBlock) {
            level.playSound(
                player,
                player!!.x,
                player.y,
                player.z,
                SoundEvents.AMETHYST_BLOCK_FALL,
                SoundSource.PLAYERS,
                0.2f,
                0.5f
            )
            return InteractionResult.FAIL
        }
        if (tag.contains("1")) {
            for (key in tag.allKeys) {
                if (key.matches(".*\\d.*".toRegex())) {
                    val currSel = NbtUtils.readBlockPos(tag.getCompound(key))
                    if (currSel == clicked) {
                        player!!.cooldowns.addCooldown(this, 20)
                        level.playSound(
                            player,
                            player.x,
                            player.y,
                            player.z,
                            SoundEvents.CHAIN_BREAK,
                            SoundSource.PLAYERS,
                            1.0f,
                            1.0f
                        )
                        tag.remove("First")
                        return InteractionResult.FAIL
                    }
                }
            }
            if (player!!.isShiftKeyDown) {
                val currSel = NbtUtils.readBlockPos(tag.getCompound("1"))
                println(currSel.toString() + "Is Connected to" + clicked)
                player.cooldowns.addCooldown(this, 10)
                level.playSound(
                    player,
                    player.x,
                    player.y,
                    player.z,
                    SoundEvents.CHAIN_PLACE,
                    SoundSource.PLAYERS,
                    1.0f,
                    1.0f
                )
                return InteractionResult.CONSUME
            }
            val last = tag.allKeys.size
            tag.put((last + 1).toString(), NbtUtils.writeBlockPos(clicked))
            pContext.itemInHand.tag = tag
            player.cooldowns.addCooldown(this, 10)
            level.playSound(
                player,
                player.x,
                player.y,
                player.z,
                SoundEvents.IRON_TRAPDOOR_OPEN,
                SoundSource.PLAYERS,
                1.0f,
                1.0f
            )
            return InteractionResult.SUCCESS
        }
        tag.put("1", NbtUtils.writeBlockPos(clicked))
        pContext.itemInHand.tag = tag
        player!!.cooldowns.addCooldown(this, 10)
        level.playSound(
            player,
            player.x,
            player.y,
            player.z,
            SoundEvents.IRON_TRAPDOOR_CLOSE,
            SoundSource.PLAYERS,
            1.0f,
            1.0f
        )
        return InteractionResult.SUCCESS
    }
}
